lib/active_merchant/billing/gateways/wompi.rb

```
@@ -0,0 +1,246 @@
+      end
+
+      private
+
+      def add_customer_data(post, options)
+        post[:customer_data] = {}
+        post[:customer_data][:phone_number] = options[:customer]&.[](:mob_phone)
+        post[:customer_data][:full_name] = options[:customer]&.[](:full_name)
+        post[:customer_email] = options[:customer]&.[](:email)
+        post[:redirect_url] = options[:redirect_url]
+      end
+
+      def add_address(post, creditcard, options)
+        post[:shipping_address] = add_address_data(options)
+      end
+
+      def add_address_data(options)
+        address = {}
+
+        if options[:address].present?
+          address[:address_line_1] = options[:address][:address1]
+          address[:address_line_2] = options[:address][:address2]
+          address[:city] = options[:address][:city]
+          address[:region] = options[:address][:state]
+          address[:name] = options[:customer]&.[](:full_name)
+          address[:phone_number] = options[:customer]&.[](:mob_phone)
+          address[:postal_code] = options[:address][:postal_code]
+          address[:country] = options[:address][:country]
+        end
+
+        address
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount_in_cents] = amount(money).to_i
+        post[:currency] = (options[:currency] || currency(money))
+        post[:reference] = options[:reference]
+      end
+
+      def add_payment(post, payment, options)
+        post[:payment_method] = {}
+        post[:payment_method][:type] = "CARD"
+        post[:payment_method][:token] = options[:token]
+        post[:payment_method][:installments] = options[:installments] || 1
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def headers(action)
+        key = reference_in_params?(action) ? :private_key : :public_key
+
+        {
+          'accept' => '*/*',
+          'Content-Type' => 'application/json',
+          'Authorization' => "Bearer #{@options.send(:[], key)}"
```

Should this fail fast if there's no auth token?